npm_and_yarn/lib/dependabot/npm_and_yarn.rb

```diff
@@ -25,3 +25,111 @@
+      "YN0075" => { message: "Prolog instantiation error", error_class: Dependabot::PrologInstantiationError },
+      "YN0077" => { message: "Ghost architecture", error_class: Dependabot::GhostArchitecture },
+      "YN0080" => { message: "Network disabled", error_class: Dependabot::NetworkDisabled },
+      "YN0081" => { message: "Network unsafe HTTP", error_class: Dependabot::NetworkUnsafeHTTP }
+    }.freeze, T::Hash[String, { message: String, error_class: T.class_of(Dependabot::DependabotError) }])
+
+    # Used to check if package manager registry is public npm registry
+    NPM_REGISTRY = "registry.npmjs.org"
+
+    # Used to check if url is http or https
+    HTTP_CHECK_REGEX = %r{https?://}
+
+    # Error message when a package.json name include invalid characters
+    INVALID_NAME_IN_PACKAGE_JSON = "Name contains illegal characters"
+
+    # Used to identify error messages indicating a package is missing, unreachable,
+    # or there are network issues (e.g., ENOBUFS, ETIMEDOUT, registry down).
+    PACKAGE_MISSING_REGEX = /(ENOBUFS|ETIMEDOUT|The registry may be down)/
+
+    # Used to check if error message contains timeout fetching package
+    TIMEOUT_FETCHING_PACKAGE_REGEX = %r{(?<url>.+)/(?<package>[^/]+): ETIMEDOUT}
+
+    # Used to identify git unreachable error
+    UNREACHABLE_GIT_CHECK_REGEX = /ls-remote --tags --heads (?<url>.*)/
+
+    # Used to check if yarn workspace is enabled in non-private workspace
+    ONLY_PRIVATE_WORKSPACE_TEXT = "Workspaces can only be enabled in priva"
+
+    # Used to identify local path error in yarn when installing sub-dependency
+    SUB_DEP_LOCAL_PATH_TEXT = "refers to a non-existing file"
+
+    # Used to identify invalid package error when package is not found in registry
+    INVALID_PACKAGE_REGEX = /Can't add "(?<package_req>.*)": invalid/
+
+    # Used to identify error if node_modules state file not resolved
+    NODE_MODULES_STATE_FILE_NOT_FOUND = "Couldn't find the node_modules state file"
+
+    # Used to find error message in yarn error output
+    YARN_USAGE_ERROR_TEXT = "Usage Error:"
+
+    # Used to identify error if tarball is not in network
+    TARBALL_IS_NOT_IN_NETWORK = "Tarball is not in network and can not be located in cache"
+
+    # Finding errors such as "The current Node version 20.15.1 does not satisfy the required version 20.11.0"
+    NODE_VERSION_NOT_SATISFY_REGEX = /The current .*Node.* version.*does not satisfy the required version/
+
+    # Used to identify if authentication failure error
+    AUTHENTICATION_TOKEN_NOT_PROVIDED = "authentication token not provided"
+    AUTHENTICATION_IS_NOT_CONFIGURED = "No authentication configured for request"
+
+    # Used to identify if error message is related to yarn workspaces
+    DEPENDENCY_CONFLICT = "conflicts with direct dependency"
+
+    # Group of patterns to validate error message and raise specific error
+    VALIDATION_GROUP_PATTERNS = T.let([
```

Comment:
Validation group patterns are used to manage custom error handlings to keep the business logic cleaner.